### DISCUSSION TITLE
Twig Variables/Data in Pimcore 11

### DISCUSSION THREAD

**Discussion by limenet** [pimcore/pimcore#16407] [Q&A / Support] [answered]
While upgrading a project from Pimcore 10 to 11 I encountered an unexpected behavior when passing variables to the Twig templates.

---

**Pimcore 10**

Up until Pimcore 10, we'd write our controller actions like this:

```php
/**
 * @Template
 */
public function someAction(SomeService $someService): array
{
    return [
        'data' => $someService->getData()
    ];
}
```

And in the Twig template we'd have access to `data`, e.g.

```twig
{{ data }}
```

This worked no matter if we specified the template in the annotation, used the auto-discovery feature (controller + action), or specified/overwrote the template in the Pimcore Document.

---

**Pimcore 11**

As template auto-discovery is no longer supported in Pimcore 11, we switched to attributes:

```php
#[Template('some.html.twig')]
public function someAction(SomeService $someService): array
{
    return [
        'data' => $someService->getData()
    ];
}
```

Now, however, the behavior is different based on whether the template is specified/overwritten on the `Document` instance:

- `$document->getTemplate() === null`
  - Behavior remains unchanged and `{{ data }}` works.
  - Symfony logic in `\Symfony\Bridge\Twig\EventListener\TemplateAttributeListener` is used
- `$document->getTemplate() !== null`
  - We no longer have access to `data` inside our Twig template
  - `someService` is now available in the Twig template
  - Pimcore logic in `\Pimcore\Bundle\CoreBundle\EventListener\Frontend\ContentTemplateListener` is used

---

To restore the old behavior, the following patch can be applied:

```patch
--- a/bundles/CoreBundle/src/EventListener/Frontend/ContentTemplateListener.php
+++ b/bundles/CoreBundle/src/EventListener/Frontend/ContentTemplateListener.php
@@ -58,6 +58,7 @@
     public function onKernelView(ViewEvent $event): void
     {
         $request = $event->getRequest();
+        $parameters = $event->getControllerResult();

         if (!$this->matchesPimcoreContext($request, PimcoreContextResolver::CONTEXT_DEFAULT)) {
             return;
@@ -70,7 +71,7 @@
             return;
         }

-        $parameters = $this->resolveParameters($event->controllerArgumentsEvent, $attribute?->vars);
+        $parameters ??= $this->resolveParameters($event->controllerArgumentsEvent, $attribute?->vars);
         $status = 200;

         if (interface_exists('Symfony\\Component\\Form\\FormInterface')) {
```
---

**Question**: Is this new behavior intentional or am I missing something?

**Comment by fashxp** [comment]
hmm, not sure if that is intentional ...

**Reply by limenet** [reply, marked as answer]
Since it is not linked or referenced automatically in this discussion: https://github.com/pimcore/pimcore/issues/16474 was created and the behavior is fixed via https://github.com/pimcore/pimcore/pull/16456.
